src/class/cdc/cdc_device.c

```diff
@@ -402,10 +402,10 @@ bool cdcd_xfer_cb(uint8_t rhport, uint8_t ep_addr, xfer_result_t result, uint32_
   // Data sent to host, we could continue to fetch data tx fifo to send.
   // But it will cause incorrect baudrate set in line coding.
   // Though maybe the baudrate is not really important !!!
-//  if ( ep_addr == p_cdc->ep_in )
-//  {
-//
-//  }
+  if ( ep_addr == p_cdc->ep_in )
+  {
+    if ( xferred_bytes && (0 == (xferred_bytes % CFG_TUD_CDC_EPSIZE)) ) usbd_edpt_xfer(TUD_OPT_RHPORT, p_cdc->ep_in, NULL, 0);

```

Comment:
This fixes the problem of endpoint-size transmission that is not perceived on host side due to missing zero length packet.
But as @pigrew noticed zero length packets will be send in more cases than needed.

Changing condition to:
```if ( xferred_bytes && (0 == (xferred_bytes % CFG_TUD_CDC_EPSIZE)) && p_cdc->tx_ff.count == 0)```
...could skip sending ZLP when there is something more in fifo already and will be sent in next flush.

Reply:
Thanks @kasjer ; since this is a now dated branch, I've merged it with the presently master branch, and have implemented your suggestion.

Reply:
@hathach anything is missing here or is it possible to merge it. I'm also interested in that patch. Thanks

Reply:
I lack the motivation, since the current implementation works well and has no issue with any driver at all. Although this PR is the right implementation, however since it doesn't "fix" any problems. I am reluctant to test & merge.

Reply:
AFAIK, it's unfortunately class-specific whether a ZLP is sent if a transfer size is a multiple of the packet size.

The USB spec itself punts on the issue, saying "the two sides must have agreed on how much data is to be sent ahead of time. When that total amount of data is sent, _or any packet that is less than max packet size is sent_, the transfer ends immediately. Sending more data than agreed upon ahead of time is undefined."

This definition allows a transfer to end on a packet multiple _without_ sending a subsequent ZLP, since the two sides could've agreed ahead of time that "`64`, `128`, `192`, etc bytes will be the full transfer, we'll both stop after that". This is the behavior the [control pipe](https://github.com/hathach/tinyusb/issues/139#issuecomment-530954118) uses when the device can send the host _all_ its requested data.

I think the above definition _also_ allows the ZLP case. But I couldn't tell you which device classes do this. What does the CDC spec say on this?

Reply:
> I lack the motivation, since the current implementation works well and has no issue with any driver at all. Although this PR is the right implementation, however since it doesn't "fix" any problems. I am reluctant to test & merge.

I thought that I provided a concise demonstration back on Jan 4 that the current implementation does not work when writing 64 characters with tud_cdc_write_char():

https://github.com/hathach/tinyusb/pull/260#issue-359242869

Reply:
> > I lack the motivation, since the current implementation works well and has no issue with any driver at all. Although this PR is the right implementation, however since it doesn't "fix" any problems. I am reluctant to test & merge.
> 
> I thought that I provided a concise demonstration back on Jan 4 that the current implementation does not work when writing 64 characters with tud_cdc_write_char():
> 
> [#260 (comment)](https://github.com/hathach/tinyusb/pull/260#issue-359242869)

aha, how could I miss this. I possibly was debugging something else while reading this. I will test it out